src/js/protocols/stm32usbdfu.js

```diff
@@ -559,59 +600,52 @@ STM32DFU_protocol.prototype.upload_procedure = function (step) {
-                                    self.upload_procedure(1);
-                                });
-                            });
-                        }
                     } else if (typeof chipInfo.external_flash !== "undefined") {
-                        // external flash, flash to the 3rd partition.
+                        // external flash
+                        nextAction = 2; // no option bytes
+
                         self.chipInfo = chipInfo;
                         self.flash_layout = chipInfo.external_flash;
+                    } else {
+                        console.log('Failed to detect internal or external flash');
+                        self.cleanup();
+                    }
+
+                    if (typeof nextAction !== "undefined") {
+                        GUI.log(i18n.getMessage('dfu_device_flash_info', (self.flash_layout.total_size / 1024).toString()));
 
-                        var firmware_partition_index = 2;
-                        var firmware_sectors = self.flash_layout.sectors[firmware_partition_index];
-                        var firmware_partition_size = firmware_sectors.total_size;
+                        // verify all addresses in the hex are writable.
 
-                        self.available_flash_size = firmware_partition_size;
+                        const unusableBlocks = [];
 
-                        GUI.log(i18n.getMessage('dfu_device_flash_info', (self.flash_layout.total_size / 1024).toString()));
+                        for (const block of self.hex.data) {
+                            const usable = self.isBlockUsable(block.address, block.bytes);
+                            if (!usable) {
+                                unusableBlocks.push(block);
+                            }
+                        }
 
-                        if (self.hex.bytes_total > self.available_flash_size) {
-                            GUI.log(i18n.getMessage('dfu_error_image_size',
-                                [(self.hex.bytes_total / 1024.0).toFixed(1),
-                                (self.available_flash_size / 1024.0).toFixed(1)]));
+                        if (unusableBlocks.length > 0) {
+                            GUI.log(i18n.getMessage('dfu_hex_address_errors'));
                             self.cleanup();
                         } else {
                             self.getFunctionalDescriptor(0, function (descriptor, resultCode) {
                                 self.transferSize = resultCode ? 2048 : descriptor.wTransferSize;
                                 console.log('Using transfer size: ' + self.transferSize);
                                 self.clearStatus(function () {
-                                    self.upload_procedure(2); // no option bytes to deal with
+                                    self.upload_procedure(nextAction);
                                 });
                             });
                         }
-                    } else {
```

Comment:
Why remove this exception?

Reply:
that's odd, I don't recall doing that, well spotted, I will re-instate.

Reply:
Ahh, I see what happened, it moved upwards to line 619, see above.